storage/pebble/payloads.go

```diff
@@ -36,67 +37,84 @@ func NewPayloads(db *badger.DB, index *Index, guarantees *Guarantees, seals *Sea
 	return p
 }
 
-func (p *Payloads) storeTx(blockID flow.Identifier, payload *flow.Payload) func(*transaction.Tx) error {
+func (p *Payloads) storeTx(blockID flow.Identifier, payload *flow.Payload) func(storage.PebbleReaderBatchWriter) error {
 	// For correct payloads, the execution result is part of the payload or it's already stored
 	// in storage. If execution result is not present in either of those places, we error.
 	// ATTENTION: this is unnecessarily complex if we have execution receipt which points an execution result
 	// which is not included in current payload but was incorporated in one of previous blocks.
 
-	return func(tx *transaction.Tx) error {
-
+	return func(rw storage.PebbleReaderBatchWriter) error {
 		resultsByID := payload.Results.Lookup()
 		fullReceipts := make([]*flow.ExecutionReceipt, 0, len(payload.Receipts))
 		var err error
+		batch := rw.IndexedBatch()
 		for _, meta := range payload.Receipts {
 			result, ok := resultsByID[meta.ResultID]
 			if !ok {
-				result, err = p.results.ByIDTx(meta.ResultID)(tx)
-				if err != nil {
-					if errors.Is(err, storage.ErrNotFound) {
+				// if result is not in the payload of the current block,
+				// it should be in either storage or previous blocks.
+				// reading from the indexed batch can read the block from previous block
+				result, err = p.results.byID(meta.ResultID)(batch)
+				if errors.Is(err, storage.ErrNotFound) {
+					// if the result is not in the previous blocks, check storage
+					result, err = p.results.ByID(meta.ResultID)
```

Comment:
When storing payloads, we require the receipt must refer to a result that exists. Given multiple receipts could refer to the same result, and we want the same result only be stored once, this makes this query a bit tricky: when verifying if the result has been stored, not long we need to look at the write-set in the current batch from previous writes, but also the database. 

Badger does this two lookup automatically in a transaction, but pebble requires us to do these two queries separately.